buildbot/master/files/config/factories.py

```diff
@@ -104,6 +105,106 @@ def make_step(self, command):
         return step_class(**step_kwargs)
 
 
+class StepsYAMLParsingStep(buildstep.ShellMixin, buildstep.BuildStep):
+    """
+    Step which resolves the in-tree YAML and dynamically adds test steps.
+    """
+
+    haltOnFailure = True
+    flunkOnFailure = True
+
+    def __init__(self, builder_name, environment, yaml_path, **kwargs):
+        kwargs = self.setupShellMixin(kwargs)
+        buildstep.BuildStep.__init__(self, **kwargs)
+        self.builder_name = builder_name
+        self.environment = environment
+        self.yaml_path = yaml_path
+
+    @defer.inlineCallbacks
+    def run(self):
+        try:  # Prevent the buildbot master from dying. Will remove after test.
+            try:
+                print_yaml_cmd = "cat {}".format(self.yaml_path)
+                cmd = yield self.makeRemoteShellCommand(
+                    command=[print_yaml_cmd],
+                    collectStdout=True
+                )
+                yield self.runCommand(cmd)
+
+                result = cmd.results()
+                if result != util.SUCCESS:
+                    raise Exception(
+                        "Command failed with return code: {}" + str(cmd.rc)
+                    )
+                else:
+                    builder_steps = yaml.safe_load(cmd.stdout)
+                    commands = builder_steps[self.builder_name]
+                    dynamic_steps = [self.make_step(command)
+                                     for command in commands]
+            except Exception as e:  # Bad step configuration, fail build
+                # Capture the exception and re-raise with a friendly message
+                raise Exception("Bad configuration, " +
+                                "unable to convert to steps" +
+                                str(e))
+
+            # TODO: windows compatibility (use a custom script for this?)
+            pkill_step = steps.ShellCommand(command=["pkill", "-x", "servo"],
+                                            decodeRC={0: SUCCESS, 1: SUCCESS})
+            static_steps = [pkill_step]
+
+            self.build.steps += static_steps + dynamic_steps
+
+            defer.returnValue(result)
+        except Exception as e:
+            print(str(e))
```

Comment:
Actually, the entire outermost try/except block isn't helpful and we don't need it, because Buildbot handles exceptions safely (e.g. http://build.servo.org/builders/mac-rel-css/builds/1606). Please remove it, sorry for the go-around.